### ISSUE TITLE
aws_elastic_beanstalk_environment Error: Provider produced inconsistent final plan

### ISSUE THREAD

**Issue by yashshanker** [hashicorp/terraform-provider-aws#13823]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
```
Terraform v0.12.26
+ provider.aws v2.66.0
+ provider.random v2.2.1
+ provider.time v0.5.0
```

### Affected Resource(s)

* aws_elastic_beanstalk_environment

### Terraform Configuration Files

```hcl
resource "aws_elastic_beanstalk_environment" "my-eb-env" {
  name        = "${aws_elastic_beanstalk_application.my-eb-app.name}-v1"
  application = aws_elastic_beanstalk_application.my-eb-app.name

  solution_stack_name = "64bit Amazon Linux 2018.03 v2.20.3 running Multi-container Docker 19.03.6-ce (Generic)"
  tier                = "WebServer"

  setting {
    namespace = "aws:autoscaling:launchconfiguration"
    name      = "ImageId"
    value     = var.ami
  }

  # Application configuration variables
  setting {
    namespace = "aws:elasticbeanstalk:application:environment"
    name      = "READ_DB_HOST"
    value     = aws_db_instance.my_replica_db.address
  }
}
```

### Panic Output

```hcl
Error: Provider produced inconsistent final plan

When expanding the plan for aws_elastic_beanstalk_environment.my-eb-env to include
new values learned so far during apply, provider "registry.terraform.io/-/aws"
produced an invalid new value for .setting: planned set element
cty.ObjectVal(map[string]cty.Value{"name":cty.StringVal("READ_DB_HOST"),
"namespace":cty.StringVal("aws:elasticbeanstalk:application:environment"),
"resource":cty.NullVal(cty.String), "value":cty.StringVal("something")}) does not
correlate with any element in actual.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

### Expected Behavior

This is a part of a bigger configuration, which has all my infra's configuration. When I do `terraform apply` for the first time, it applies successfully and sets up my infra. Then subsequently, if I change anything, suppose security groups' ingress/egress CIDRs, it should apply and make those changes.

### Actual Behavior

The first `terraform apply` get applied successfully, but subsequently, when I do changes, suppose changes in security groups' ingress/egress CIDRs, it ends up with an error which is mentioned above.

By the way, there are more settings in my elastic_beanstalk_environment resource, and this error is for all of them.

### Steps to Reproduce

1. `terraform apply`

### References

* <a href="https://github.com/hashicorp/terraform/issues/25287" target="_blank">The above described issue has started coming in after I added this to my configuration. But I am not really sure they are related.</a>

**Comment by vanwalj** [comment]
Just got a similar issue

```
Error: Provider produced inconsistent final plan
 When expanding the plan for aws_s3_bucket_object.lambda to include new values
 learned so far during apply, provider "registry.terraform.io/hashicorp/aws"
 changed the planned action from CreateThenDelete to DeleteThenCreate.
 This is a bug in the provider, which should be reported in the provider's own
 issue tracker.
```

Everything was working fine so far, but I changed a value: `scaling_configuration`.`seconds_until_auto_pause` in an `aws_rds_cluster` ressource, an since then, all subsequent apply failed

**Comment by yashshanker** [comment]
@vanwalj if you are sure it is because of some change in rds cluster, then most probably there is something wrong with the config.

I just saw that I was writing the wrong configuration for RDS [here](https://github.com/hashicorp/terraform/issues/25287), and once I resolved my config for RDS, there were no more errors in other resources. 

Try it out, and let me know.

**Comment by vanwalj** [comment]
@yashshanker just found what caused the issue, we were applying our terraform config in a ci pipeline using the docker `hashicorp/terraform:full` image which was recently updated to use the "0.13-beta" version of terraform which literally obliterate our infrastructure, rolling back to 12.26 fixed the issue

**Comment by mslipper** [comment]
I have a reproducible test case for this.

First, create an SQS queue resource with a corresponding `data` resource as follows:

```hcl
resource "aws_sqs_queue" "queue" {
  name = "super_duper_queue"
  max_message_size = 2048
  visibility_timeout_seconds = 300
}

data "aws_sqs_queue" "reports" {
  name = aws_sqs_queue.queue.name
  depends_on = [aws_sqs_queue.reports]
}
```

Then, pipe the report URL into an EB environment variable:

```hcl
// other directives cut for brevity
  setting {
    namespace = "aws:elasticbeanstalk:application:environment"
    name = "SQS_QUEUE_URL"
    value = data.aws_sqs_queue.queue.url
  }
```

It'll work on the first apply, but will stop working after that with the following error:

```
Error: Provider produced inconsistent final plan
When expanding the plan for aws_elastic_beanstalk_environment.my_env to
include new values learned so far during apply, provider
"registry.terraform.io/-/aws" produced an invalid new value for .setting:
planned set element
cty.ObjectVal(map[string]cty.Value{"name":cty.StringVal("InstanceTypes"),
"namespace":cty.StringVal("aws:ec2:instances"),
"resource":cty.NullVal(cty.String), "value":cty.StringVal("t2.micro")}) does
not correlate with any element in actual.
This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

Commenting out the `depends_on` in the `data` resource fixes the issue.

**Comment by zygimantas** [comment]
The same happens while changing cpu from 1024 to 256 in `aws_ecs_task_definition`

```
When expanding the plan for module.aws_region_us1.aws_ecs_task_definition.main
to include new values learned so far during apply, provider
"registry.terraform.io/hashicorp/aws" changed the planned action from
CreateThenDelete to DeleteThenCreate.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

Terraform v0.13.0
+ provider registry.terraform.io/-/aws v3.3.0
+ provider registry.terraform.io/hashicorp/aws v3.3.0

**Comment by bflad** [comment]
For folks running into the `changed the planned action from CreateThenDelete to DeleteThenCreate` error,  I believe this was fixed upstream in Terraform CLI by https://github.com/hashicorp/terraform/pull/26192, which was released in Terraform 0.13.3. If you are still running into this same error after upgrading Terraform CLI itself, please open an issue upstream in https://github.com/hashicorp/terraform/issues since that particular error cannot be fixed in Terraform Providers (see referenced pull request for more details).
